core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,401 @@
+            } else {
+              assert(taskInfo.successful)
+              // The first attempt of this task should have failed since it was killed. Thus,
+              // either the task attempt or the stage attempt number should be more than 0.
+              assert(taskInfo.attemptNumber > 0 || taskEnd.stageAttemptId > 0)
+            }
+          } else {
+            // The second task is never touched and thus should succeed on the first attempt.
+            assert(taskInfo.successful)
+            assert(taskInfo.attemptNumber === 0)
+          }
+        }
+      }
+      sc.addSparkListener(listener)
+      val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex((pid, iter) => {
+        val sleepTimeSeconds = if (pid == 0) 1 else 10
+        Thread.sleep(sleepTimeSeconds * 1000L)
+        List.fill(pid + 1)(pid * 2 + 1).iterator
+      }, preservesPartitioning = true).repartition(1).sum()
+      assert(jobResult > 1)
+      // 4 tasks: 2 from first stage, one retry due to decom, one more in the second stage.
+      val tasksSeen = listener.getTasksFinished(10.seconds)
+      assert(tasksSeen.size >= 4, s"Expected at least 4 tasks but got $tasksSeen")
+    } finally {
+      ss.close()
+    }
+  }
+
+  test("decommission workers ensure that fetch failures lead to rerun") {
+    val conf = appConf
+    conf.set(config.Tests.TEST_NO_STAGE_RETRY, false)
+    conf.set(config.UNREGISTER_OUTPUT_ON_HOST_ON_FETCH_FAILURE, true)
+    makeWorkers(2)
+    sc = createSparkContext(conf)
+    val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+    val executorToDecom = executorIdToWorkerInfo.keysIterator.next
+    val workerToDecomInfo = executorIdToWorkerInfo(executorToDecom)
+    val workerToDecomHost = workerToDecomInfo.host
+    val workerToDecomPort = workerToDecomInfo.port
+    // The setup of this job is similar to the one above: 2 stage job with first stage having
+    // long and short tasks. Except that we want the shuffle output to be regenerated on a
+    // fetch failure instead of an executor lost. Since it is hard to "trigger a fetch failure",
+    // we manually raise the FetchFailed exception when the 2nd stage's task runs and require that
+    // fetch failure to trigger a recomputation.
+    logInfo(s"Will try to decom the task running on executor $executorToDecom")
+    val listener = new RootStageAwareListener {
+      override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+        val taskInfo = taskEnd.taskInfo
+        if (taskInfo.executorId == executorToDecom && taskInfo.attemptNumber == 0 &&
+          taskEnd.stageAttemptId == 0) {
+          decommissionWorkerOnMaster(workerToDecomInfo,
+            "decommission worker after task on it is done")
+        }
+      }
+    }
+    sc.addSparkListener(listener)
+    val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex((index, _) => {
```

Comment:
Nit: 
```suggestion
    val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex { (index, _) =>
```

Reply:
This repeats at several other places.

Style guide I use:
https://github.com/databricks/scala-style-guide#anonymous-methods

Reply:
Here in this exact line I am very much wrong as there is second param. 


But:
- https://github.com/apache/spark/blob/2ebfaccd7710e4152ccb73abfe91ecddf694356d/core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala#L115

-https://github.com/apache/spark/blob/2ebfaccd7710e4152ccb73abfe91ecddf694356d/core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala#L284